File: api/src/main/java/hirez/api/StatusPage.java
package hirez.api;

import hirez.api.object.statuspage.*;
import io.reactivex.rxjava3.core.Single;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class StatusPage {
    private final Endpoint endpoint;

    public Single<Summary> getSummary() {
        return endpoint.get(Summary.class, combineUrl("/summary.json"));
    }

    public Single<StatusData> getStatus() {
        return endpoint.get(StatusData.class, combineUrl("/status.json"));
    }

    public Single<Components> getComponents() {
        return endpoint.get(Components.class, combineUrl("/components.json"));
    }

    public Single<Incidents> getUnresolvedIncidents() {
        return endpoint.get(Incidents.class, combineUrl("/incidents/unresolved.json"));
    }

    public Single<Incidents> getIncidents() {
        return endpoint.get(Incidents.class, combineUrl("/incidents.json"));
    }

    public Single<ScheduledMaintenances> getUpcomingScheduledMaintenances() {
        return endpoint.get(ScheduledMaintenances.class, combineUrl("/scheduled-maintenances/upcoming.json"));
    }

    public Single<ScheduledMaintenances> getActiveScheduledMaintenances() {
        return endpoint.get(ScheduledMaintenances.class, combineUrl("/scheduled-maintenances/active.json"));
    }

    public Single<ScheduledMaintenances> getAllScheduledMaintenances() {
        return endpoint.get(ScheduledMaintenances.class, combineUrl("/scheduled-maintenances.json"));
    }

    private String combineUrl(String endpoint) {
        return "http://stk4xr7r1y0r.statuspage.io/api/v2" + endpoint;
    }
}

File: api/src/main/kotlin/hirez/statuspage/StatusPage.kt
package hirez.statuspage

import com.google.gson.Gson
import hirez.api.Http
import hirez.json.statuspage.Components
import hirez.json.statuspage.Incidents
import hirez.json.statuspage.ScheduledMaintenances
import hirez.json.statuspage.StatusData
import hirez.json.statuspage.Summary

/**
 * @author Damian Staszewski [damian@stachuofficial.tv]
 * @version %I%, %G%
 * @since 1.0
 */
class StatusPage(private val http: Http) {
	
	private fun combineUrl(endpoint: String): String = "http://stk4xr7r1y0r.statuspage.io/api/v2$endpoint"
	
	val summary
		get() = http.call<Summary>(combineUrl("/summary.json"))
	
	val status
		get() = http.call<StatusData>(combineUrl("/status.json"))
	
	val components
		get() = http.call<Components>(combineUrl("/components.json"))
	
	val unresolvedIncidents
		get() = http.call<Incidents>(combineUrl("/incidents/unresolved.json"))
	val incidents
		get() = http.call<Incidents>(combineUrl("/incidents.json"))
	
	val upcomingScheduledMaintenances
		get() = http.call<ScheduledMaintenances>(combineUrl("/scheduled-maintenances/upcoming.json"))
	val activeScheduledMaintenances
		get() = http.call<ScheduledMaintenances>(combineUrl("/scheduled-maintenances/active.json"))
	val allScheduledMaintenances
		get() = http.call<ScheduledMaintenances>(combineUrl("/scheduled-maintenances.json"))
}
